indy_ledger_response_parser/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+
+        let (revoc_reg_def_id, revoc_reg, timestamp) = match reply.result() {
+            GetRevocRegReplyResult::GetRevocRegReplyResultV0(res) => (res.revoc_reg_def_id, res.data, res.txn_time),
+            GetRevocRegReplyResult::GetRevocRegReplyResultV1(res) => (
+                res.txn.data.revoc_reg_def_id,
+                res.txn.data.value,
+                res.txn_metadata.creation_time,
+            ),
+        };
+
+        Ok(RevocationRegistryInfo {
+            revoc_reg: RevocationRegistry::RevocationRegistryV1(revoc_reg),
+            revoc_reg_def_id,
+            timestamp,
+        })
+    }
+
+    pub fn parse_get_revoc_reg_delta_response(
+        &self,
+        get_revoc_reg_delta_response: &str,
+    ) -> IndyResult<RevocationRegistryDeltaInfo> {
+        let reply: Reply<GetRevocRegDeltaReplyResult> = Self::parse_response(get_revoc_reg_delta_response)?;
+
+        let (revoc_reg_def_id, revoc_reg) = match reply.result() {
+            GetRevocRegDeltaReplyResult::GetRevocRegDeltaReplyResultV0(res) => (res.revoc_reg_def_id, res.data),
+            GetRevocRegDeltaReplyResult::GetRevocRegDeltaReplyResultV1(res) => {
+                (res.txn.data.revoc_reg_def_id, res.txn.data.value)
+            }
+        };
+
+        let revoc_reg_delta = RevocationRegistryDeltaV1 {
+            value: json!(UrsaRevocationDelta::from_parts(
+                revoc_reg.value.accum_from.map(|accum| accum.value).as_ref(),
+                &revoc_reg.value.accum_to.value,
+                &revoc_reg.value.issued,
+                &revoc_reg.value.revoked,
+            )),
+        };
+
+        Ok(RevocationRegistryDeltaInfo {
+            revoc_reg_delta: RevocationRegistryDelta::RevocationRegistryDeltaV1(revoc_reg_delta),
+            revoc_reg_def_id,
+            timestamp: revoc_reg.value.accum_to.txn_time,
+        })
+    }
+
+    pub fn parse_response<T>(response: &str) -> IndyResult<Reply<T>>
+    where
+        T: DeserializeOwned + ReplyType + ::std::fmt::Debug,
+    {
+        let message: serde_json::Value =
+            serde_json::from_str(&response).to_indy(IndyErrorKind::InvalidTransaction, "Response is invalid json")?;
+
+        if message["op"] == json!("REPLY") && message["result"]["type"] != json!(T::get_type()) {
+            return Err(err_msg(IndyErrorKind::InvalidTransaction, "Invalid response type"));
+        }
+
+        let message: Message<T> = serde_json::from_value(message).to_indy(
+            IndyErrorKind::LedgerItemNotFound,
+            "Structure doesn't correspond to type. Most probably not found",
+        )?; // FIXME: Review how we handle not found
+
+        match message {
+            Message::Reject(response) | Message::ReqNACK(response) => Err(err_msg(
+                IndyErrorKind::InvalidTransaction,
+                format!("Transaction has been failed: {:?}", response.reason),
+            )),
+            Message::Reply(reply) => Ok(reply),
+        }
+    }
```

Comment:
I agree that this piece is an ugly workaround. Your suggestion is good in spirit, but it needed a little bit of tweaking to work, as it incorrectly assumes that `type` is a sibling field of `op`, but is in fact part of `Reply` result. To be specific, this is a simplified version of a response:
```
{
  "op": "REPLY",
  "result": {
     ...
    "type": "107"
  }
}
```
(The part about `#[serde(try_from = "MsgWithReplyType")]` on `Message` seems to contradict the rest of your suggestion, where you are proposing to deserialize into `MsgWithTypedReply`, and then convert directly into `Reply`. Moreover, assuming you mean `MsgWithTypedReply`, it would also take generic parameters to account for. But I assume it is no longer relevant.)

Reply:
Aaaahhh, my bad! It was obvious but I missed that. It can still be done in the same fashion though, but it becomes a bit annoying because of the manually specified bounds. I'll close this and start a new convo to make the suggestion.